pkg/api/vm/handler.go

```diff
@@ -626,6 +641,116 @@ func (h *vmActionHandler) removeVolume(ctx context.Context, namespace, name stri
 		Error()
 }
 
+// cloneVM creates a VM which uses volume cloning from the source VM.
+func (h *vmActionHandler) cloneVM(name string, namespace string, input CloneInput) error {
+	vm, err := h.vmCache.Get(namespace, name)
+	if err != nil {
+		return fmt.Errorf("cannot get vm %s/%s, err: %w", namespace, name, err)
+	}
+	newVM := &kubevirtv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        input.TargetVM,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+		},
+		Spec: *vm.Spec.DeepCopy(),
+	}
+
+	newPVCs, secretNameMap, err := h.cloneVolumes(newVM)
+	if err != nil {
+		return fmt.Errorf("clone volumes error for new vm %s/%s, err %w", newVM.Namespace, newVM.Name, err)
+	}
+	newPVCsString, err := json.Marshal(newPVCs)
+	if err != nil {
+		return fmt.Errorf("cannot marshal value %+v, err: %w", newPVCs, err)
+	}
+
+	newVM.ObjectMeta.Annotations[util.AnnotationVolumeClaimTemplates] = string(newPVCsString)
+	if newVM, err = h.vms.Create(newVM); err != nil {
+		return fmt.Errorf("cannot create newVM %+v, err: %w", newVM, err)
+	}
+
+	for oldSecretName, newSecretName := range secretNameMap {
+		secret, err := h.secretCache.Get(namespace, oldSecretName)
+		if err != nil {
+			return fmt.Errorf("cannot get secret %s/%s, err: %w", namespace, name, err)
+		}
+
+		newSecret := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      newSecretName,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion: newVM.APIVersion,
+						Kind:       newVM.Kind,
+						Name:       newVM.Name,
+						UID:        newVM.UID,
+					},
+				},
+			},
+			Data:       secret.Data,
+			StringData: secret.StringData,
+			Type:       secret.Type,
+		}
+		if _, err = h.secretClient.Create(&newSecret); err != nil {
+			return fmt.Errorf("cannot create secret %+v, err: %w", newSecret, err)
```

Comment:
Maybe better not print secret contents out to the error message? Guess a secret is sufficient?

Reply:
Fixed it. Thank you.